### PR TITLE
refactor(status): update status.condition in DSC and DSCI

### PR DESCRIPTION
**Pull request by zdtsw**
## Description
re-wrok on https://github.com/opendatahub-io/opendatahub-operator/pull/977

feat(status): update status.condition in DSC and DSCI
    
  for Condition:
  - create a generic function FailedComponentCondition() for all components to use, return condition and error
  - keep status types to only "avaiable" "progressing" degraded" and "reconcilesuccess"
  - status:ReconcileComplete is renamed to ReconcileSuccess
  - status:upgradeable is removed
  
  for Phase: Created, Ready, Error, deleting
  - set phase to Error if Argo CRD exist
  - remove Progressing to use Created instead
  
  chore:
  - fix typo and some missing customized error message
  - fix Message when created 

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/status/component_condition.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Component init condition.
+// type: <component>Ready
+// status: Unknown
+// reason: ReconcileStart
+func SetInitComponentCondition(componentName string, enabled bool) conditionsv1.Condition {
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/status/component_condition.go]: I update the name from SetInitComponentCondition() to InitControllerCondition()
and a bunch of others: removed Set in prefix

main idea is:
in controller: 

-   on DSC/DSCI/FTer condition

| case    | to call |
| -------- | ------- |
|DSC/DSCI just created|UpdateCondition(oldcondition, InitControllerCondition())|
|deleting DSC by configmap|UpdateCondition(oldcondition, UnavailableCondition())|
| DSC exist but no DSCI |UpdateCondition(oldcondition, UnavailableCondition())|
| DSC/DSCI success reconciled| SetCompleteCondition()|
| FTer success applied| SetCompleteCondition()|
| DSCI failed reconcile|SetErrorCondition(...,CapabilityFailed,...)|
| DSC failed reconcile| SetErrorCondition()|
| FTer failed to create|SetErrorCondition()|

-   on component condition (under DSC): 

| result    | to call |
| -------- | ------- |
| success + removed|  RemoveComponentCondition()|
|failure| UpdateCondition(oldcondition, reconcileCondition)|
|success  + managed| UpdateCondition(oldcondition, reconcileCondition) |
| sepcial on argo| UpdateCondition(oldcondition, ArgoExistCondition()) |
| component just set to managed|UpdateCondition(oldcondition, InitComponentCondition())|

in each component logic: 
| result    | return |
| -------- | ------- |
|success|SuccessComponentCondition(...) used as reconcileCondition |
|failure|FailedComponentCondition(....)  used as reconcileCondition|

File: controllers/status/component_condition.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+// Component reconcile failed.
+// type: <component>Ready
+// status: False
+// reason: ReconcileFailed
+// message: <derive from err>
+func UpdateFailedCondition(componentName string, err error) (conditionsv1.Condition, error) {
+	FailedCondition := setFailedComponentCondition(componentName)
+	FailedCondition.Message = err.Error()
+	return FailedCondition, err
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/status/component_condition.go:46]: to convert err into String and set as condition.Message.

File: controllers/status/controller_condition.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	return &conditionsv1.Condition{
+		Type:    conditionsv1.ConditionAvailable,
+		Status:  corev1.ConditionFalse,
+		Reason:  reason,
+		Message: message,
+	}
+}
+
+// UpdateCondition is top caller in DSC controller to handle different case: init, check, error, successetc.
+func UpdateCondition(conditions *[]conditionsv1.Condition, newCondition conditionsv1.Condition) {
```
[the remark this answers is not in the document]
Comment (ykaliuta) [reply controllers/status/controller_condition.go:96]: So it should have been called `UpdateComponentCondition`. But I did not change my opinion that both wrappers are misleading. They assume there is some logic  related to component behind the call while they just operate on the provided lists. The intention (the list and the provided member of the list) is clear from the call itself without the wrapper.

File: controllers/status/component_condition.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ykaliuta) [reply controllers/status/component_condition.go:46]: It is the point of the function, but it does not answer my question :)

File: controllers/status/component_condition.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ykaliuta) [reply controllers/status/component_condition.go]: So, why not New* when you just create a structure? It makes intentions clear, while reading "InitComponentCondition" at the first glance I would read is "Initialize ComponentCondition".

[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/status/component_condition.go]: i take your suggestion and renamed it also skipped UpdateCondition on this init creation phase

**Pull request closed without merging**